Python/context.c

```diff
@@ -257,6 +263,37 @@ PyContext_Exit(PyObject *octx)
 }
 
 
+void
+_PyContext_ExitThreadOwned(PyThreadState *ts)
+{
+    assert(ts != NULL);
+    // notify_context_watchers requires the notification to come from the
+    // affected thread, so we can only exit the context(s) if ts belongs to the
+    // current thread.
+    _Bool on_thread = ts == _PyThreadState_GET();
```

Comment:
I dislike this design of only doing certain things if the thread state is attached, because thread states can switch around for the same thread (e.g. with interpreter switching). I would rather swapping to the passed thread state and restoring the old one.

Reply:
I refactored a bit to always exit the thread-owned contexts.  Now the only difference is it skips emitting a `Py_CONTEXT_SWITCHED` event if the calling thread does not own the tstate.